docs/notes/shutdown.txt

```diff
@@ -0,0 +1,68 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+
+
+qdrouterd shutdown procedure
+============================
+
+
+How does the router shut down when interrupted with ctrl-C ?
+
+
+  1. ctrl-c is caught by server.c::handle().
+     it comes packaged by proton as PN_PROACTOR_INTERRUPT.
+     handle() sends it back through proton 
+     with pn_proactor_interrupt()
+     to interrupt the next thread, and then returns false.
+
+
+  2. The false return from handle() is received by thread_run().
+     If there is a container, thread_run() calls 
+     qd_conn_event_batch_complete() to let the container know 
+     that the work batch is finished.
+
+      2.1 qd_conn_event_batch_complete() loops through the 
+          free_link_session_list for this container. 
+          If the connection is not already closed, it frees 
+          each session's pn_link and pn_session, if they exist.
+
+
+  3. Once all threads have exited from thread_run(), 
+     qd_server_run() logs a "Shut Down" message, and returns to 
+     main_process() in main.c .
+
+
+  4. main_process() sets to NULL the dispatch pointer that 
+     everybody was using, and then frees the structure.
```

Comment:
> and then frees the structure

This is the most "interesting" part. Do tell more!

Reply:
You think there should be more detail here?

Reply:
Well, yeah. `qd_dispatch_free` does a lot, and it calls functions that do a lot.

Reply:
OK, will do.  Oops. I just assumed it was trivial.  Bad assumption.